src/array/cpu/rowwise_pick.h

```diff
@@ -92,68 +94,99 @@ COOMatrix CSRRowWisePick(CSRMatrix mat, IdArray rows,
   //
   // [02/29/2020 update]: OMP is disabled for now since batch-wise parallelism is more
   //   significant. (minjie)
-  IdArray picked_row = Full(-1, num_rows * num_picks, sizeof(IdxType) * 8, ctx);
-  IdArray picked_col = Full(-1, num_rows * num_picks, sizeof(IdxType) * 8, ctx);
-  IdArray picked_idx = Full(-1, num_rows * num_picks, sizeof(IdxType) * 8, ctx);
+  IdArray picked_row = NDArray::Empty({num_rows * num_picks},
+                                      DLDataType{kDLInt, 8*sizeof(IdxType), 1},
+                                      ctx);
+  IdArray picked_col = NDArray::Empty({num_rows * num_picks},
+                                      DLDataType{kDLInt, 8*sizeof(IdxType), 1},
+                                      ctx);
+  IdArray picked_idx = NDArray::Empty({num_rows * num_picks},
+                                      DLDataType{kDLInt, 8*sizeof(IdxType), 1},
+                                      ctx);
   IdxType* picked_rdata = static_cast<IdxType*>(picked_row->data);
   IdxType* picked_cdata = static_cast<IdxType*>(picked_col->data);
   IdxType* picked_idata = static_cast<IdxType*>(picked_idx->data);
 
-  bool all_has_fanout = true;
-#pragma omp parallel for reduction(&&:all_has_fanout)
-  for (int64_t i = 0; i < num_rows; ++i) {
-    const IdxType rid = rows_data[i];
-    const IdxType len = indptr[rid + 1] - indptr[rid];
-    // If a node has no neighbor then all_has_fanout must be false even if replace is
-    // true.
-    all_has_fanout = all_has_fanout && (len >= (replace ? 1 : num_picks));
-  }
+  const int num_threads = omp_get_max_threads();
+  std::vector<int64_t> global_prefix(num_threads+1, 0);
 
-#pragma omp parallel for
-  for (int64_t i = 0; i < num_rows; ++i) {
-    const IdxType rid = rows_data[i];
-    CHECK_LT(rid, mat.num_rows);
-    const IdxType off = indptr[rid];
-    const IdxType len = indptr[rid + 1] - off;
-    if (len == 0)
-      continue;
+#pragma omp parallel num_threads(num_threads)
+  {
+    const int thread_id = omp_get_thread_num();
 
-    if (len <= num_picks && !replace) {
-      // nnz <= num_picks and w/o replacement, take all nnz
-      for (int64_t j = 0; j < len; ++j) {
-        picked_rdata[i * num_picks + j] = rid;
-        picked_cdata[i * num_picks + j] = indices[off + j];
-        picked_idata[i * num_picks + j] = data? data[off + j] : off + j;
+    const int64_t start_i = thread_id * (num_rows/num_threads) +
+        std::min(static_cast<int64_t>(thread_id), num_rows % num_threads);
+    const int64_t end_i = (thread_id + 1) * (num_rows/num_threads) +
+        std::min(static_cast<int64_t>(thread_id + 1), num_rows % num_threads);
+    assert(thread_id + 1 < num_threads || end_i == num_rows);
+
+    const int64_t num_local = end_i - start_i;
+
+    // make sure we don't have to pay initialization cost
+    std::unique_ptr<int64_t[]> local_prefix(new int64_t[num_local + 1]);
```

Comment:
Dumb question.  Why do we wrap around this with a unique pointer?

Reply:
This is just to ensure it gets free'd via RAII. The reason we don't use a `std::vector<int64_t>` is to avoid paying the cost of zeroing all of the memory.